---
title: "Species_openness"
author: "Zane Kliesmete"
date: "2023-08-17"
output: pdf
editor_options: 
  chunk_output_type: inline
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(ggplot2)
libs<-c("ggplot2","ggpubr","RMySQL","tidyverse","cowplot","data.table","GenomicRanges", "patchwork", "broom", "DESeq2", "figpatch", "ape", "ggtree")

sapply(libs, require, character.only=T)
source("../ATACseq/scripts/TFBS_plotting_functions_ATAC.R")

tissueColors <- c( "#9E0142" ,"#D53E4F", "#F46D43", "#FDAE61", "#FFD92F" ,"#ABDDA4", "#66C2A5", "#3288BD", "#5E4FA2")
names(tissueColors) <-  c("adrenal gland", "brain", "heart", "kidney", "large intestine", "lung", "muscle", "stomach", "thymus")

specificityColors <- c( "#A3753B", "#CC9B57", "#E7CF97", "#F8EDD0", "#F7F7F7", "#D2EEEA", "#99D7CE", "#5DACA5", "#33847E")
names(specificityColors) <-  c(1:9)

regionColors2 <- c("#9CA578","#3288BD","#E37063")
names(regionColors2) <-  c("Enhancer", "Promoter","Expression")

regionColors <- c("#9CA578","#3288BD","#E37063")
names(regionColors) <-  c("Enhancer", "Promoter","Expression")

basic_theme<-  theme(axis.title=element_text(size=8.5),
        axis.text = element_text(color="black", size=7),
        #legend.title = element_text(size = 8), 
        legend.text = element_text(size = 6.5),
        legend.key.size = unit(0.8, "lines"),
        legend.margin=margin(0.3,0.3,0.3,0.3),
        legend.box.margin=margin(0.2,0.2,0.2,0.2))


theme_set( theme_bw( base_size = 8 ) +
             theme(legend.title = element_blank(),
                   legend.key.size = unit(3,"mm")) )

basic_theme_ins<-  theme(axis.title=element_text(size=9.5),
                     axis.text = element_text(color="black", size=8.5),
                     legend.title = element_text(size = 7.5), 
                     legend.text = element_text(size = 8),
                     legend.key.size = unit(0.5, "lines"),
                     legend.margin=margin(0.1,0.1,0.1,0.1),
                     legend.box.margin=margin(0.1,0.1,0.1,0.1),
                     panel.grid.major = element_line(size = 0.35),
                     panel.grid.minor = element_blank(),
                     strip.text.x = element_text(size = 10))

basic_theme_ins2<-  theme(axis.title=element_text(size=8),
                         axis.text = element_text(color="black", size=7),
                         legend.title = element_text(size = 7.3), 
                         legend.text = element_text(size = 7),
                         legend.key.size = unit(0.5, "lines"),
                         legend.margin=margin(0,0,0,0),
                         legend.box.margin=margin(0,0,0,0),
                         legend.background = element_blank(), 
                         legend.box.background = element_blank(),
                         panel.grid.major = element_line(size = 0.35),
                         panel.grid.minor = element_blank(),
                         strip.text = element_text(size = 8))


theme_ines<-theme_bw( base_size = 8 ) +
             theme(legend.title = element_blank(),
                   legend.position = "bottom",
                   legend.box = "horizontal",
                   legend.direction = "horizontal",
                   legend.key.size = unit(3,"mm"),
                   axis.text = element_text(color="black"))

```

# Focus on NPCs

## ATAC-seq peak presence / absence

```{r pp}

stringent_set<-readRDS("../ATACseq/RDS/jammPeaks_vs_ATACPeaks_NPC_stringent.rds") %>%
  dplyr::mutate(openness=gsub("only","specific",openness),
                openness=ifelse(openness=="Always open", "Conserved", openness))

peakOL_plot_enhancer_NPCs <- stringent_set %>% 
  dplyr::count(total, openness, assignment) %>%
  group_by(assignment,total) %>% 
  mutate(frac = n/sum(n),
         openness = factor(openness, levels=c("Not open","Macaque-specific","Human-specific","Conserved")),
         assignment = ifelse(assignment == "enh","Enhancer","Promoter")) %>%
  filter(assignment == "Enhancer") %>% 
  ggplot(aes(x=as.factor(total), y=n, fill = openness))+
  geom_bar(position="fill", stat="identity")+
  scale_fill_manual(values=c("#F4F4F6","#D1D1DC","#8B8BA7","#69698C")) +
  ylab("Fraction of Enhancers")+
  xlab("PD")

peakOL_plot_promoter_NPCs <- stringent_set %>%
  dplyr::count(total, openness, assignment) %>%
  group_by(assignment,total) %>% 
  mutate(frac = n/sum(n),
         openness = factor(openness, levels=c("Not open","Macaque-specific","Human-specific","Conserved")),
         assignment = ifelse(assignment == "enh","Enhancer","Promoter")) %>%
  filter(assignment == "Promoter") %>% 
  ggplot(aes(x=as.factor(total), y=n, fill = openness))+
  geom_bar(position="fill", stat="identity")+
  scale_fill_manual(values=c("#F4F4F6","#D1D1DC","#8B8BA7","#69698C")) +
  ylab("Fraction of Promoters")+
  xlab("PD")

peakOL_plot_enhancer_NPCs|peakOL_plot_promoter_NPCs


# both together in a facet_grid setting
peakOL_plot_NPCs <- stringent_set %>% 
  dplyr::count(total, openness, assignment) %>%
  group_by(assignment,total) %>% 
  mutate(frac = n/sum(n),
         openness = factor(openness, levels=c("Not open","Macaque-specific","Human-specific","Conserved")),
         assignment = ifelse(assignment == "enh","Enhancer","Promoter")) %>%
  ggplot(aes(x=as.factor(total), y=n, fill = openness))+
  geom_bar(position="fill", stat="identity")+
  scale_fill_manual(values=c("#F4F4F6","#D1D1DC","#8B8BA7","#69698C")) +
  ylab("Fraction of CREs")+
  xlab("PD")+
  basic_theme_ins2+
  theme(legend.title=element_blank())+
  facet_grid(assignment~.)+
  guides(fill = guide_legend(nrow = 2))

```

## LFCs of DA and DE

```{r GetLFCs}

DHS_to_gene <- readRDS("../CRE_to_Gene/DHS_to_gene.rds") %>%
  dplyr::group_by(region_id, gene_id, total, CGI, assignment) %>% 
  dplyr::mutate(distance = min(distance)) %>% 
  distinct(region_id, gene_id, total, CGI, assignment, distance)

jamm_forATAC_NPC_all<-readRDS("../ATACseq/RDS/jammPeaks_vs_ATACPeaks_NPC_all.rds")


atacDDS <- readRDS("../ATACseq/RDS/atacDDS.rds")
DA<-lapply( c("NPC","iPSC"), function(i){
    tmp <- colData(atacDDS)
    tmp <- tmp[tmp$cell_type == i,]
    tmpdds <- DESeqDataSetFromMatrix( colData = tmp,
                                      countData = DESeq2::counts(atacDDS)[,rownames(tmp)],
                                      design = ~ species)
    tmpdds <-DESeq(tmpdds)
    results(tmpdds) %>% data.frame %>% rownames_to_column("region_id") %>% 
      mutate(celltype = i) %>% 
      separate_wider_delim(region_id, ".", names = c(NA, "region_id")) %>% 
      mutate(region_id = as.numeric(region_id)) %>% 
      inner_join(DHS_to_gene)
}) %>% bind_rows

exprdds <- readRDS("../RNAseq/RDS/dds_clean.rds")
DE<-lapply( c("NPC","iPSC"), function(i){
  tmp <- colData(exprdds)
  tmp <- tmp[tmp$Differentiation == i,]
  tmpdds <- DESeqDataSetFromMatrix( colData = tmp,
                                    countData = DESeq2::counts(exprdds)[,rownames(tmp)],
                                    design = ~ Species)
  tmpdds <-DESeq(tmpdds)
  results(tmpdds) %>% data.frame %>% rownames_to_column("gene_id") %>% 
    mutate(celltype = i)
}) %>% bind_rows


comb.lfc<- DE %>% left_join(DA, by=c("gene_id","celltype"), suffix=c("",".da"))

saveRDS(comb.lfc, "../RNAseq/RDS/DE_DA_table.rds")

expr_pleio<- readRDS("../roadmap_expression_summaries/summarized_expression_allTissues.rds") %>% 
  group_by(gene_id) %>% 
  summarise( expr_pleio = length(tissue),
             expr_pleio = factor(expr_pleio, levels=1:9 )) 

comb.lfc <- comb.lfc %>% left_join(expr_pleio)

boot_ci<- function(x,n=100){
  y <- sapply(1:n, function(i){ 
    median( sample(x, size=length(x), replace = T), na.rm = T) }) 
  sort(y)[ c(floor(0.05*n),ceiling(0.95*n)) ]
}

expr_LFC<-comb.lfc %>% 
  distinct( PD=expr_pleio, gene_id, celltype, log2FoldChange, baseMean, padj) %>% 
  drop_na() %>%  group_by(celltype,PD) %>% 
  reframe( lfc =median(abs(log2FoldChange)),
           level = median(baseMean),
           ci = boot_ci(abs(log2FoldChange)) ,
           ci_level = boot_ci(baseMean),
           cidir= c("low","high"),
           nDE = sum(padj<=0.1),
           notDE = sum(padj>0.1),
           type ="expression") %>% 
  pivot_wider(values_from = c(ci,ci_level), names_from = cidir)

      
cre_LFC<-comb.lfc %>% 
  distinct( PD= as.factor(total), region_id,type=assignment, 
            celltype, log2FoldChange.da, baseMean.da, padj.da) %>% 
  filter(region_id %in% jamm_forATAC_NPC_all$region_id[jamm_forATAC_NPC_all$openness!="Not open"]) %>%
  drop_na() %>%  group_by(celltype,PD,type) %>% 
  reframe( lfc =median(abs(log2FoldChange.da)),
           level = median(baseMean.da),
           ci = boot_ci(abs(log2FoldChange.da)) ,
           ci_level = boot_ci(baseMean.da),
           cidir=c("low","high"),
           nDE = sum(padj.da<=0.1),
           notDE = sum(padj.da>0.1)) %>% 
    pivot_wider(values_from = c(ci,ci_level), names_from = cidir)

LFC_sum <- bind_rows( expr_LFC, cre_LFC)

```

# Compare LFCs ----------------------------------------------------
```{r plotLFCs}
lfc_npc_summary<-LFC_sum %>% mutate(PD = as.numeric(PD),
                   type = case_when( type == "prom" ~ "Promoter",
                                     type == "enh" ~ "Enhancer",
                                     type == "expression" ~ "Expression")) %>% 
  filter(celltype == "NPC") %>% 
  ggplot(aes(x=PD,y=lfc,ymin=ci_low,ymax=ci_high,col=type)) +
  geom_line()+
  geom_pointrange(size=0.3)+
  scale_x_continuous(breaks=1:9)+
  ylab(expression(paste("|",log[2](frac(a[H], a[M])),"|"))) +
  scale_color_manual(values = regionColors2)+
  basic_theme_ins2+
  theme(legend.position = c(0.8,0.91), legend.title=element_blank())


level_npc_summary<-LFC_sum %>% mutate(PD = as.numeric(PD),
                                      type = case_when( type == "prom" ~ "Promoter",
                                                        type == "enh" ~ "Enhancer",
                                                        type == "expression" ~ "Expression")) %>% 
  filter(celltype == "NPC") %>% 
  ggplot(aes(x=PD,y=level,ymin=ci_level_low,ymax=ci_level_low,col=type)) +
  geom_line()+
  geom_pointrange(size=1/2)+
  scale_x_continuous(breaks=1:9)+
  scale_y_log10()+
  ylab(expression(log[2](bar(a))))+
  scale_color_manual(values = regionColors)

```

## Odds ratios, Fisher's test

```{r cor}

ft<- comb.lfc %>% 
  filter(region_id %in% jamm_forATAC_NPC_all$region_id[jamm_forATAC_NPC_all$openness!="Not open"]) %>%
  group_by(celltype, assignment, gene_id, total) %>% 
  summarise(DE = sum(abs(log2FoldChange)>=1, padj<=0.1,na.rm=T)>0,
            DA = sum(padj.da<=0.1,na.rm=T),
            lfc = max(abs(log2FoldChange)),
            lfc.da.sum = sum(abs(log2FoldChange.da)),
            lfc.da.max = max(abs(log2FoldChange.da)),
            n=length(total)) %>% 
  group_by(gene_id,celltype,assignment) %>% 
  mutate( hasDA_prom = sum( assignment == "prom" & DA>0)>0,
          hasDA_enh  = sum( assignment == "enh"  & DA>0)>0,
          hasDA_PD9_prom = sum( assignment == "prom"  & DA>0 & total==9)>0,
          hasDA_PD9_enh  = sum( assignment == "enh"  & DA>0 & total==9)>0,
          hasDA_PDM_prom    = sum( assignment == "prom"  & DA>0 & total>1 & total<9)>0,
          hasDA_PDM_enh    = sum( assignment == "enh"  & DA>0 & total>1 & total<9)>0,
          hasDA_PD1_prom    = sum( assignment == "prom"  & DA>0 & total==1 )>0,
          hasDA_PD1_enh    = sum( assignment == "enh"  & DA>0 & total==1 )>0)

x<-ft %>% filter(celltype == "NPC",hasDA_enh==F)
cor.test(x$lfc.da.max,x$lfc, 
         method = "spearman")

# are DE genes enriched if a gene has at least one DA promoter
ft %>% ungroup %>% filter(celltype =="NPC" , hasDA_enh==F) %>% 
  distinct(gene_id,DE,hasDA_prom) %>% 
  dplyr::count(DE,hasDA_prom) %>% 
  summarise( broom::tidy(fisher.test( matrix(n,2,2))))


# are DE genes enriched if a gene has at least one DA enhancer
ft %>% ungroup %>% filter(celltype =="NPC" , hasDA_prom==F) %>% 
  distinct(gene_id,DE,hasDA_enh) %>% 
  dplyr::count(DE,hasDA_enh) %>% 
  summarise( broom::tidy(fisher.test( matrix(n,2,2))))

```


```{r matchedF}

df<- comb.lfc %>% 
  filter(celltype == "NPC",
         region_id %in% jamm_forATAC_NPC_all$region_id[jamm_forATAC_NPC_all$openness!="Not open"]) %>% 
  mutate(DA = ifelse(padj.da<0.1, 1, 0),
         DE = ifelse(padj<0.1 & abs(log2FoldChange)>1, 1, 0),
         DE = ifelse(is.na(DE),0, DE),
         PD_cat = case_when(total %in% c(1) ~ "specific",
                            total %in% c(9) ~ "pleiotropic",
                            T ~ "intermediate")) 

gene_profiles<-df %>% 
  group_by(gene_id, DA, PD_cat, assignment) %>% 
  dplyr::summarise(nn = length(gene_id)) %>% 
  pivot_wider(names_from = c(PD_cat, assignment, DA), values_from = nn) %>% 
  replace(is.na(.), 0) 

```


```{r evalnot, eval=F}

# running this takes a while, you can just also load it below line 373
fisher_list<-list()
for (ass in c("enh","prom")){
  for (PD in c(1,2,9)){
    
    CRE_combi_of_interest<-df %>% filter(assignment == ass, total == PD) %>%
      mutate(combi = paste(PD_cat, assignment, sep = "_")) %>% slice_head(n=1) %>% pull(combi)
    print(CRE_combi_of_interest)
    
    # Select rows where combi_1 =>1 (DE)
    selected_rows <- gene_profiles %>%
      filter(!!rlang::sym(glue::glue("{CRE_combi_of_interest}_1")) > 0)
    
    # Filter matching rows, i.e. genes with non-DE +1 and DE -1
    matching_rows <- selected_rows %>%
      mutate(!!rlang::sym(glue::glue("{CRE_combi_of_interest}_1")) := !!rlang::sym(glue::glue("{CRE_combi_of_interest}_1")) - 1,
             !!rlang::sym(glue::glue("{CRE_combi_of_interest}_0")) := !!rlang::sym(glue::glue("{CRE_combi_of_interest}_0")) + 1)
  
    leftover_rows <- gene_profiles %>%
      filter(!gene_id %in% matching_rows$gene_id)
    matching_rows$gene_id_matched<-NA
    
    for (r in 1:nrow(matching_rows)){
      suppressMessages({
      joined<-inner_join(leftover_rows, matching_rows[r,2:(ncol(matching_rows)-1)])
      })
      if (dim(joined)[1]>0){
        matching_rows$gene_id_matched[r]<-joined$gene_id[1]
        leftover_rows<-leftover_rows %>% filter(gene_id !=joined$gene_id[1])
      }
    }
    
    matching_rows_clean<-matching_rows %>% drop_na()
    
    combi<-bind_rows(
      selected_rows %>% mutate(DA_target = 1) %>% 
        filter(gene_id %in% matching_rows_clean$gene_id),
      matching_rows_clean %>% 
        mutate(DA_target = 0) %>% ungroup() %>% 
        dplyr::select(-gene_id) %>% 
        dplyr::rename(gene_id = gene_id_matched)
    ) %>% 
      left_join(df %>% distinct(gene_id, DE)) %>% 
      group_by(DA_target, DE) %>% 
      dplyr::summarise(nn = length(DA_target))
    
    fisher_list[[paste0(PD,"_",ass)]]<-
      fisher.test(matrix(combi$nn, 2,2, byrow = T)) %>% 
      broom::tidy() %>% 
      dplyr::mutate(contrast = CRE_combi_of_interest, 
                    total = PD, 
                    assignment = ass, 
                    n_per_DA_group = dim(matching_rows_clean)[1])
    
  }
}

fishers_table<-bind_rows(fisher_list, .id="combi") %>% 
  mutate( PD  = factor(total), 
          type = ifelse(assignment == "enh","Enhancer","Promoter"),
          padj = p.adjust(p.value, method = "BH"),
          sig.code = case_when( padj <= 1e-3 ~ "***",
                                padj <= 1e-2 ~ "**",
                                padj <= 0.05 ~ "*",
                                T ~ "")) 

saveRDS(fishers_table, "ATACseq/RDS/DA_DE_tables/matched_fishers_table_1M9.rds")
```

```{r run}
# Matched fishers table
fishers_table<-readRDS("../ATACseq/RDS/DA_DE_tables/matched_fishers_table_1M9.rds")

fig3_fishtab<-fishers_table  %>% 
  distinct(estimate, .keep_all=TRUE) %>% 
  separate(contrast, into = c("PD_cat", "assignment"), sep = "_") %>% 
  mutate(PD_cat2 = case_when(PD_cat == "specific" ~ "1",
                             PD_cat == "intermediate" ~ "2-8",
                             PD_cat == "pleiotropic" ~ "9")) %>% 
  ggplot(aes(x=PD_cat2, y=estimate,ymin=conf.low, 
             ymax=conf.high, 
             label=sig.code,
             color=type)) +
  geom_hline(yintercept = 1,linetype=2, color="darkgrey")+
  geom_pointrange(position=position_dodge(width=0.5), size=0.3)+ 
  geom_text(aes(y=conf.high+0.5),
            size=3, position = position_dodge(width=0.5), show.legend=FALSE)+
  ylab("Odds Ratio (DA vs. DE)")+
  scale_y_log10()+
  scale_color_manual(values=regionColors)+
  basic_theme_ins2+
  theme(axis.text = element_text(color="black"),
        legend.position=c(0.18, 0.91),
        legend.title = element_blank())+
  xlab("Grouped PD")

```


### New figure 3

```{r roller}

treeLengths<-readRDS("../Roller2021/PhyloTreeLengths.rds")

evoage<-ggplot(treeLengths, aes(x=as.factor(total), y=rel_tree_length, fill=as.factor(total)))+
  geom_boxplot(notch = T, outlier.size = 0.1)+
  theme_bw()+
  xlab("PD")+
  ylab(expression(paste("Activity conservation (", lambda[i]/lambda, ")")))+
  scale_fill_manual(values=specificityColors)+
  basic_theme_ins2+
  theme(legend.position = "none",
        axis.text = element_text(color="black"))


mammaltree<-read.tree("/data/share/htp/TRNP1/paper_data/Co-evolution-TRNP1-and-GI/protein/trees/mammaltree.txt")

speciesTree<-drop.tip(mammaltree, mammaltree$tip.label[!mammaltree$tip.label %in% c("Macaca_mulatta","Callithrix_jacchus","Felis_catus","Canis_lupus", "Equus_ferus", "Mus_musculus", "Rattus_norvegicus", "Oryctolagus_cuniculus", "Sus_scrofa", "Monodelphis_domestica")])

# also plot species tree
speciesTree_plot<-speciesTree
speciesTree_plot$tip.label<- case_when(speciesTree_plot$tip.label == "Macaca_mulatta" ~ "Macaque",
                                       speciesTree_plot$tip.label == "Callithrix_jacchus" ~ "Marmoset",
                                       speciesTree_plot$tip.label == "Felis_catus" ~ "Cat",
                                       speciesTree_plot$tip.label == "Canis_lupus" ~ "Dog", 
                                       speciesTree_plot$tip.label == "Equus_ferus" ~ "Horse", 
                                       speciesTree_plot$tip.label == "Mus_musculus" ~ "Mouse", 
                                       speciesTree_plot$tip.label == "Rattus_norvegicus" ~ "Rat",
                                       speciesTree_plot$tip.label == "Oryctolagus_cuniculus" ~ "Rabbit", 
                                       speciesTree_plot$tip.label == "Sus_scrofa" ~ "Pig", 
                                       speciesTree_plot$tip.label == "Monodelphis_domestica" ~ "Opossum")

mamtree<-ggtree(speciesTree_plot, lwd=0.3)+
  geom_tiplab(fontface="italic", size=2, geom = "text")+ 
  xlim(0,200) +
  theme(plot.margin = unit(c(0, 0, 0,0), "cm"))+
  geom_treescale(width=20, x=0, y=0.05, offset=0.15, fontsize=2, linesize=0.35)

```


```{r new3}

# Construct plot 3
fig2A<-fig("../../figures/npc_peaks_pleio_base.pdf", 
            b_margin = ggplot2::margin(-10,0,-10,-10), link_dim = T)+
  theme(plot.margin = margin(0,0,0,0))

fig3_top<-plot_grid(fig2A+theme(axis.text = element_text(color="black")),
                    NULL, peakOL_plot_NPCs+theme(legend.position="top"), 
                    rel_widths=c(1,0.06,1.34), ncol=3,
                    scale=c(0.9,1, 1), labels=c("A","","B"), 
                    label_size=12, label_x = c(0,-0.05))

fig3_middle<-plot_grid(mamtree,evoage, rel_widths = c(1,1.3), scale = c(0.8, 1), labels=c("C","D"), label_size=12, label_x = c(0,-0.05))

fig3_bottom<-plot_grid(lfc_npc_summary+theme(legend.position=c(0.75,0.87)),
                       NULL, fig3_fishtab, ncol=3, rel_widths=c(1.1,0.1,1),
                       labels=c("E","F"), label_size=12, label_x = c(0,0))
                

plot_grid(fig3_top, NULL, fig3_middle,NULL, fig3_bottom, ncol=1, rel_heights=c(1,0.05,0.85,0.05,0.85))
ggsave(file="../../figures/Fig3.pdf",width =162.5*0.75, height=160, units = "mm")

```




### Supplementary figure
Some heatmaps and odds

```{r odds_npcs}

atacDDS <- readRDS("../ATACseq/RDS/atacDDS.rds")
vsdMat<-assay(varianceStabilizingTransformation(atacDDS))
vsdMat2<-vsdMat[,grepl("NPC", colnames(vsdMat))]
# Heatmap for most variable regions
topVar <- order(rowVars(vsdMat2), decreasing = T)[1:1000]
topVarMat <- vsdMat2[topVar, ]

# Plot heatmap
ann_colors = list(
  species = c("Human"="#8B8BA7","Macaque"="#D1D1DC")
)

library(pheatmap)
pdf("../figures/DA_heatmap.pdf", height = 4.5, width = 5)
pheatmap(topVarMat,
         cluster_rows = TRUE,
         show_rownames = FALSE,
         show_colnames = FALSE,
         annotation_col = dplyr::transmute(as.data.frame(colData(atacDDS)[grepl("NPC", rownames(colData(atacDDS))),]), species=ifelse(species=="cynomolgus", "Macaque", "Human")),
         annotation_colors=ann_colors,
         cutree_rows = 1,
         cutree_cols = 1,
         scale = "none")
dev.off()



exprdds <- readRDS("../expression_conservation/RDS/dds_clean.rds")
vsdMat<-assay(varianceStabilizingTransformation(exprdds))
vsdMat2<-vsdMat[,grepl("d5|d7|d9", colnames(vsdMat))]
# Heatmap for most variable regions
topVar <- order(rowVars(vsdMat2), decreasing = T)[1:1000]
topVarMat <- vsdMat2[topVar, ]

pdf("../figures/DE_heatmap.pdf", height = 4.5, width = 5)
pheatmap(topVarMat,
         cluster_rows = TRUE,
         show_rownames = FALSE,
         show_colnames = FALSE,
         annotation_col = dplyr::transmute(as.data.frame(colData(exprdds)[grepl("d5|d7|d9", rownames(colData(exprdds))),]), species=ifelse(Species=="macFas", "Macaque", "Human")),
         annotation_colors=ann_colors,
         cutree_rows = 1,
         cutree_cols = 1,
         scale = "none")
dev.off()

```



## Count the number of overlapping peaks

```{r peak_clusters}

peakOL_plot_machum_NPCs <- stringent_set %>% 
  mutate(assignment = ifelse(assignment == "enh","Enhancer","Promoter")) %>% 
  filter(openness %in% c("Conserved","Macaque-specific","Human-specific")) %>% 
  group_by(assignment, total, openness) %>% 
  dplyr::summarise(n=length(total)) %>%
  ggplot(aes(x=as.factor(total), y=n, fill=openness))+
  geom_col()+#position=position_dodge())+
  scale_fill_manual(values=c("Macaque-specific"="#D1D1DC","Human-specific"="#8B8BA7","Conserved"="#69698C")) +
  facet_grid(.~assignment)+
  ylab("Number of CREs")+
  xlab("PD")


sup2A<-fig("../figures/DA_heatmap.pdf", b_margin = ggplot2::margin(-10,0,-10,-10), 
            link_dim = T)+ theme(plot.margin = margin(0,0,0,0))

sup2B<-fig("../figures/DE_heatmap.pdf", b_margin = ggplot2::margin(-10,0,-10,-10), 
            link_dim = T)+ theme(plot.margin = margin(0,0,0,0))


plot_grid(
  plot_grid(sup2A, sup2B, labels=c("A","B"), label_size=12),
  plot_grid(NULL,peakOL_plot_machum_NPCs +theme(axis.text=element_text(color="black")),
            NULL, scale = 0.95,
            rel_widths=c(0.05,1,0.05), ncol = 3),
  ncol=1, rel_heights=c(1.1,0.7), labels = c("","C"), label_size =12)

ggsave("../figures/SuplFig2.pdf", height = 7*0.7, width = 5)


```


